server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -451,7 +452,15 @@ private void executeRequest(
                 logger,
                 TraceableSearchRequestOperationsListener.create(tracer, requestSpan)
             );
-            SearchRequestContext searchRequestContext = new SearchRequestContext(requestOperationsListeners, originalSearchRequest);
+            SearchRequestContext searchRequestContext = new SearchRequestContext(requestOperationsListeners, originalSearchRequest, () -> {
+                List<String> taskResourceUsages = threadPool.getThreadContext().getResponseHeaders().get(TASK_RESOURCE_USAGE);
+                if (taskResourceUsages != null && taskResourceUsages.size() > 0) {
+                    return taskResourceUsages.get(0);
+                }
+                return null;
```

Comment:
This logic should be handled by TaskResourceTrackingService? That should be the only class aware of the headers added / removed from threadContext. Other classes should leverage those methods from TaskResourceTrackingService

Reply:
Make sense. Let me actually refactor all the interactions with resources into the `TaskResourceTrackingService`